README
NAME
    Map::Tube::GraphViz - GraphViz output for Map::Tube.

SYNOPSIS
     use Map::Tube::GraphViz;

     my $obj = Map::Tube::GraphViz->new(%params);
     $obj->graph($output_file);

METHODS
    "new(%params)"
             Constructor.

            *       "callback_edge"

                     Edge callback.
                     Default value is this:
                     sub { 
                             my ($self, $from, $to) = @_;
                             $self->{'g'}->add_edge(
                                    'from' => $from,
                                    'to' => $to,
                             );
                             return;
                     }

            *       "callback_node"

                     Node callback.
                     Default value is \&Map::Tube::GraphViz::Utils::node_color.

            *       "driver"

                     GraphViz2 driver.
                     Default value is 'neato'.

            *       "g"

                     GraphViz2 object.
                     Parameters 'g' and 'name' cannot combine.
                     Default value is this:
                     GraphViz2->new(
                            'global' => {
                                    'directed' => 0,
                            },
                            $name ? (
                                    'graph' => {
                                            'label' => $name,
                                            'labelloc' => 'top',
                                    },
                            ) : (),
                     );

            *       "name"

                     Name of map.
                     Parameters 'g' and 'name' cannot combine.
                     Default value is Map::Tube->name or undef.

            *       "output"

                     GraphViz2 output.
                     It is required.
                     Default value is 'png'.
                     Possible values are every formats supported by GraphViz2 module.
                     See L<http://www.graphviz.org/content/output-formats>.

            *       "tube"

                     Map::Tube object.
                     It is required.
                     Default value is undef.

    "graph($output_file)"
             Get graph and save it to $output_file file.
             Returns undef.

ERRORS
     new():
             Cannot load UNIVERSAL::DOES module.
             Parameter 'tube' is required.
             Parameter 'tube' must be 'Map::Tube' object.
             Parameter 'output' is required.
             Unsupported 'output' parameter '%s'.
             From Map::Tube::GraphViz::Utils::color_line():
                     No color for line '%s'.
             From Class::Utils::set_params():
                     Unknown parameter '%s'.

EXAMPLE1
     use strict;
     use warnings;

     use English;
     use Error::Pure qw(err);
     use Map::Tube::GraphViz;

     # Arguments.
     if (@ARGV < 1) {
             print STDERR "Usage: $0 metro\n";
             exit 1;
     }
     my $metro = $ARGV[0];
 
     # Object.
     my $class = 'Map::Tube::'.$metro;
     eval "require $class;";
     if ($EVAL_ERROR) {
             err "Cannot load '$class' class.",
                     'Error', $EVAL_ERROR;
     }
 
     # Metro object.
     my $tube = eval "$class->new";
     if ($EVAL_ERROR) {
             err "Cannot create object for '$class' class.",
                     'Error', $EVAL_ERROR;
     }
 
     # GraphViz object.
     my $g = Map::Tube::GraphViz->new(
             'tube' => $tube,
     );
 
     # Get graph to file.
     $g->graph($metro.'.png');

     # Print file.
     system "ls -l $metro.png";

     # Output without arguments like:
     # Usage: /tmp/SZXfa2g154 metro

     # Output with 'Berlin' argument like:
     # -rw-r--r-- 1 skim skim 1503067 Jan 27 07:24 Berlin.png

EXAMPLE2
     use strict;
     use warnings;

     use English;
     use Error::Pure qw(err);
     use Map::Tube::GraphViz;
     use Map::Tube::GraphViz::Utils qw(node_color_without_label);

     # Arguments.
     if (@ARGV < 1) {
             print STDERR "Usage: $0 metro\n";
             exit 1;
     }
     my $metro = $ARGV[0];
 
     # Object.
     my $class = 'Map::Tube::'.$metro;
     eval "require $class;";
     if ($EVAL_ERROR) {
             err "Cannot load '$class' class.",
                     'Error', $EVAL_ERROR;
     }
 
     # Metro object.
     my $tube = eval "$class->new";
     if ($EVAL_ERROR) {
             err "Cannot create object for '$class' class.",
                     'Error', $EVAL_ERROR;
     }
 
     # GraphViz object.
     my $g = Map::Tube::GraphViz->new(
             'callback_node' => \&node_color_without_label,
             'tube' => $tube,
     );
 
     # Get graph to file.
     $g->graph($metro.'.png');

     # Print file.
     system "ls -l $metro.png";

     # Output without arguments like:
     # Usage: /tmp/SZXfa2g154 metro

     # Output with 'Berlin' argument like:
     # -rw-r--r-- 1 skim skim 885928 Jan 27 07:43 Berlin.png

EXAMPLE3
     use strict;
     use warnings;

     use English;
     use Error::Pure qw(err);
     use GraphViz2;
     use Map::Tube::GraphViz;
     use Map::Tube::GraphViz::Utils qw(node_color_without_label);

     # Arguments.
     if (@ARGV < 1) {
             print STDERR "Usage: $0 metro\n";
             exit 1;
     }
     my $metro = $ARGV[0];
 
     # Object.
     my $class = 'Map::Tube::'.$metro;
     eval "require $class;";
     if ($EVAL_ERROR) {
             err "Cannot load '$class' class.",
                     'Error', $EVAL_ERROR;
     }
 
     # Metro object.
     my $tube = eval "$class->new";
     if ($EVAL_ERROR) {
             err "Cannot create object for '$class' class.",
                     'Error', $EVAL_ERROR;
     }
 
     # GraphViz object.
     my $g = Map::Tube::GraphViz->new(
             'callback_node' => \&node_color_without_label,
             'g' => GraphViz2->new(
                     'global' => {
                             'directed' => 0,
                     },
                     'graph' => {
                             'label' => $metro,
                             'labelloc' => 'top',
                             'overlap' => 0,
                     },
             ),
             'tube' => $tube,
     );
 
     # Get graph to file.
     $g->graph($metro.'.png');

     # Print file.
     system "ls -l $metro.png";

     # Output without arguments like:
     # Usage: /tmp/SZXfa2g154 metro

     # Output with 'Berlin' argument like:
     # -rw-r--r-- 1 skim skim 1212857 Jan 27 07:51 Berlin.png

EXAMPLE4
     use strict;
     use warnings;

     use English;
     use Error::Pure qw(err);
     use Map::Tube::GraphViz;
     use Map::Tube::GraphViz::Utils qw(node_color_id);

     # Arguments.
     if (@ARGV < 1) {
             print STDERR "Usage: $0 metro\n";
             exit 1;
     }
     my $metro = $ARGV[0];
 
     # Object.
     my $class = 'Map::Tube::'.$metro;
     eval "require $class;";
     if ($EVAL_ERROR) {
             err "Cannot load '$class' class.",
                     'Error', $EVAL_ERROR;
     }
 
     # Metro object.
     my $tube = eval "$class->new";
     if ($EVAL_ERROR) {
             err "Cannot create object for '$class' class.",
                     'Error', $EVAL_ERROR;
     }
 
     # GraphViz object.
     my $g = Map::Tube::GraphViz->new(
             'callback_node' => \&node_color_id,
             'tube' => $tube,
     );
 
     # Get graph to file.
     $g->graph($metro.'.png');

     # Print file.
     system "ls -l $metro.png";

     # Output without arguments like:
     # Usage: /tmp/SZXfa2g154 metro

     # Output with 'Berlin' argument like:
     # -rw-r--r-- 1 skim skim 1141071 Feb 24 08:04 Berlin.png

DEPENDENCIES
    Class::Utils, English, Error::Pure, GraphViz2, List::MoreUtils,
    Map::Tube::GraphViz::Utils, Scalar::Util.

SEE ALSO
    Map::Metro::Graph
        Map::Metro graph.

    Task::Map::Tube
        Install the Map::Tube modules.

    Task::Map::Tube::Metro
        Install the Map::Tube concrete metro modules.

REPOSITORY
    <https://github.com/michal-josef-spacek/Map-Tube-GraphViz>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

CONTRIBUTORS
    Gisbert W. Selke gws@cpan.org

LICENSE AND COPYRIGHT
    © 2014-2023 Michal Josef Špaček

    Artistic License

    BSD 2-Clause License

VERSION
    0.08



Changes
0.08
 [Common]
 - Add spaces after arguments and before return.
 - Move bugtracker to github.
 - Update Module::Install to 1.21 version.
 - Update copyright years.

0.07 2020-11-03T14:04:34+01:00
 [Common]
 - Fix Makefile.PL for new version of Perl and no default »use lib '.'«.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SYNOPSIS section in doc.
 - Rewrite SEE ALSO section to new style.
 - Rewrite to use AUTHOR tests.
 - Update Module::Install to 1.19 version.
 - Update author github username.
 - Update author name.
 - Update copyright years to 2020.

 [Map::Tube::GraphViz]
 - Add 'callback_node_id' parameter.
 - Update to new GraphViz2 module version 2.59 (changed API).

 [Map::Tube::GraphViz::Utils]
 - Add example section to doc.
 - Add cycling of colors instead error (Thanks Gisbert W. Selke - #103076).
 - Add tests for color_line() subroutine.
 - Fix subroutine name in SYNOPSIS section.
 - Fix to Map::Tube::Line with required id.

0.06 2015-02-24T08:09:07+01:00
 [Map::Tube::GraphViz]
 - Add EXAMPLE4 section in doc and examples/ex4.pl.

 [Map::Tube::GraphViz::Utils]
 - Add node_color_id() subroutine.

0.05 2015-01-29T23:42:27+01:00
 - Add tests.
 - Improve constructor error check of 'tube' parameter.
 - Update examples/ex3.pl example.

0.04 2015-01-27T07:59:41+01:00
 - Add 'g' parameter for GraphViz2 object.
 - Add EXAMPLE3 section to doc with overlapping.
 - Add images for EXAMPLE sections to doc.
 - Fix graph name.
 - Fix minimal version of Perl to 5.6.0.
 - Remove 'driver' => 'neato' from EXAMPLE sections in doc (is default).

0.03 2015-01-25T01:17:05+01:00
 - Improve CONTRIBUTORS section in doc.
 - Remove Map::Tube::GraphViz from SEE ALSO section in doc.
 - Rewrite to Map::Tube 2.72 version API (Map::Tube::Node->line).

0.02 2015-01-19T23:25:15+01:00
 - Add 'name' parameter (Thanks Gisbert W. Selke).
 - Add Map::Tube::Graph, Map::Tube::Plugin::Graph and Map::Metro::Graph to
   SEE ALSO section in doc.
 - Add Map::Tube::GraphViz, Map::Tube::Text::Table and Task::Map::Tube to
   SEE ALSO section in doc.
 - Add error check to GraphViz->run method (Thanks Gisbert W. Selke).
 - Change default 'driver' parameter to 'neato'.
 - Rewrite check of output formats (Thanks Gisbert W. Selke).
 - Update copyright years to 2015.

0.01 2014-12-17T01:20:04+01:00
 - First version


Makefile.PL
use lib '.';
use strict;
use warnings;

use English;
use inc::Module::Install;

# Definition.
abstract 'GraphViz interface to the Map::Tube.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.05;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
license 'artistic_2';
name 'Map-Tube-GraphViz';
readme_from 'GraphViz.pm';
recursive_author_tests('xt');
requires 'Class::Utils' => '0.06';
requires 'Error::Pure' => '0.21';
requires 'Exporter' => 0;
requires 'GraphViz2' => 2.59;
requires 'List::MoreUtils' => 0;
requires 'Readonly' => 0;
requires 'Scalar::Util' => 0;
if ($PERL_VERSION lt v5.10.1) {
	requires 'UNIVERSAL::DOES' => 0;
}
requires 'perl' => '5.6.0';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/Map-Tube-GraphViz/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/Map-Tube-GraphViz';
resources 'repository' => 'git://github.com/michal-josef-spacek/Map-Tube-GraphViz';
test_requires 'English' => 0;
test_requires 'Error::Pure::Utils' => 0.22;
test_requires 'File::Object' => 0.08;
test_requires 'Map::Tube::Prague' => 0;
test_requires 'Test::MockObject' => 0;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.08';

# Run.
WriteAll();


examples/ex1.pl
#!/usr/bin/env perl

# Pragmas.
use strict;
use warnings;

# Modules.
use English;
use Error::Pure qw(err);
use Map::Tube::GraphViz;

# Arguments.
if (@ARGV < 1) {
        print STDERR "Usage: $0 metro\n";
        exit 1;
}
my $metro = $ARGV[0];

# Object.
my $class = 'Map::Tube::'.$metro;
eval "require $class;";
if ($EVAL_ERROR) {
        err "Cannot load '$class' class.",
                'Error', $EVAL_ERROR;
}

# Metro object.
my $tube = eval "$class->new";
if ($EVAL_ERROR) {
        err "Cannot create object for '$class' class.",
                'Error', $EVAL_ERROR;
}

# GraphViz object.
my $g = Map::Tube::GraphViz->new(
        'driver' => 'neato',
        'tube' => $tube,
);

# Get graph to file.
$g->graph($metro.'.png');

# Print file.
system "ls -l $metro.png";

# Output without arguments like:
# Usage: /tmp/SZXfa2g154 metro

# Output with 'Berlin' argument like:
# -rw-r--r-- 1 skim skim 1503067 Jan 27 07:24 Berlin.png

examples/ex2.pl
#!/usr/bin/env perl

# Pragmas.
use strict;
use warnings;

# Modules.
use English;
use Error::Pure qw(err);
use Map::Tube::GraphViz;
use Map::Tube::GraphViz::Utils qw(node_color_without_label);

# Arguments.
if (@ARGV < 1) {
        print STDERR "Usage: $0 metro\n";
        exit 1;
}
my $metro = $ARGV[0];

# Object.
my $class = 'Map::Tube::'.$metro;
eval "require $class;";
if ($EVAL_ERROR) {
        err "Cannot load '$class' class.",
                'Error', $EVAL_ERROR;
}

# Metro object.
my $tube = eval "$class->new";
if ($EVAL_ERROR) {
        err "Cannot create object for '$class' class.",
                'Error', $EVAL_ERROR;
}

# GraphViz object.
my $g = Map::Tube::GraphViz->new(
        'callback_node' => \&node_color_without_label,
        'driver' => 'neato',
        'tube' => $tube,
);

# Get graph to file.
$g->graph($metro.'.png');

# Print file.
system "ls -l $metro.png";

# Output without arguments like:
# Usage: /tmp/SZXfa2g154 metro

# Output with 'Berlin' argument like:
# -rw-r--r-- 1 skim skim 885928 Jan 27 07:43 Berlin.png

examples/ex3.pl
#!/usr/bin/env perl

# Pragmas.
use strict;
use warnings;

# Modules.
use English;
use Error::Pure qw(err);
use GraphViz2;
use Map::Tube::GraphViz;
use Map::Tube::GraphViz::Utils qw(node_color_without_label);

# Arguments.
if (@ARGV < 1) {
        print STDERR "Usage: $0 metro\n";
        exit 1;
}
my $metro = $ARGV[0];

# Object.
my $class = 'Map::Tube::'.$metro;
eval "require $class;";
if ($EVAL_ERROR) {
        err "Cannot load '$class' class.",
                'Error', $EVAL_ERROR;
}

# Metro object.
my $tube = eval "$class->new";
if ($EVAL_ERROR) {
        err "Cannot create object for '$class' class.",
                'Error', $EVAL_ERROR;
}

# GraphViz object.
my $g = Map::Tube::GraphViz->new(
        'callback_node' => \&node_color_without_label,
        'g' => GraphViz2->new(
                'global' => {
                        'directed' => 0,
                },
                'graph' => {
                        'label' => $metro,
                        'labelloc' => 'top',
                        'overlap' => 0,
                },
        ),
        'tube' => $tube,
);

# Get graph to file.
$g->graph($metro.'.png');

# Print file.
system "ls -l $metro.png";

# Output without arguments like:
# Usage: /tmp/SZXfa2g154 metro

# Output with 'Berlin' argument like:
# -rw-r--r-- 1 skim skim 1503067 Jan 27 07:24 Berlin.png

GraphViz.pm
package Map::Tube::GraphViz;

use strict;
use warnings;

use Class::Utils qw(set_params);
use English;
use Error::Pure qw(err);
use GraphViz2;
use List::MoreUtils qw(none);
use Map::Tube::GraphViz::Utils qw(node_color);
use Scalar::Util qw(blessed);

our $VERSION = 0.08;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Edge callback.
	$self->{'callback_edge'} = sub {
		my ($self, $from, $to) = @_;
		$self->{'g'}->add_edge(
			'from' => $self->{'callback_node_id'}->($self, $from),
			'to' => $self->{'callback_node_id'}->($self, $to),
		);
		return;
	};

	# Node callback.
	$self->{'callback_node'} = \&node_color;

	# Node id callback.
	$self->{'callback_node_id'} = sub {
		my ($self, $node) = @_;
		return $node->name;
	};

	# Driver.
	$self->{'driver'} = 'neato';

	# Name of map.
	$self->{'name'} = undef;

	# GraphViz2 object.
	$self->{'g'} = undef;

	# Output.
	$self->{'output'} = 'png';

	# Map::Tube object.
	$self->{'tube'} = undef;

	# Process params.
	set_params($self, @params);

	# Check Map::Tube object.
	if (! defined $self->{'tube'}) {
		err "Parameter 'tube' is required.";
	}
	if (! defined &UNIVERSAL::DOES) {
		eval {
			require UNIVERSAL::DOES;
		};
		if ($EVAL_ERROR) {
			err 'Cannot load UNIVERSAL::DOES module.';
		}
	}
	if (! blessed($self->{'tube'})
		|| ! $self->{'tube'}->DOES('Map::Tube')) {

		err "Parameter 'tube' must be 'Map::Tube' object.";
	}

	# GraphViz2 object.
	if (defined $self->{'g'}) {
		if (defined $self->{'name'}) {
			err "Parameter 'name' cannot be used with ".
				"'g' parameter.";
		}

		# Check GraphViz2 object.
		if (! blessed($self->{'g'})
			|| ! $self->{'g'}->isa('GraphViz2')) {

			err "Parameter 'g' must be 'GraphViz2' object.";
		}
	} else {
		my $name = $self->{'name'};
		if (! defined $name) {
			$name = $self->{'tube'}->name;
		}
		$self->{'g'} = GraphViz2->new(
			'global' => {
				'directed' => 0,
			},
			$name ? (
				'graph' => {
					'label' => $name,
					'labelloc' => 'top',
				},
			) : (),
		);
	}

	# Check output format.
	if (! defined $self->{'output'}) {
		err "Parameter 'output' is required.";
	}
	if (! exists $self->{'g'}->valid_output_format->{$self->{'output'}}) {
		err "Unsupported 'output' parameter '$self->{'output'}'.";
	}

	# Object.
	return $self;
}

# Get graph.
sub graph {
	my ($self, $output_file) = @_;

	my $node_cache_hr = {};
	foreach my $node (values %{$self->{'tube'}->nodes}) {
		$self->{'callback_node'}->($self, $node);
		$node_cache_hr->{$node->id} = $node;
	}
	my @processed;
	foreach my $node (values %{$self->{'tube'}->nodes}) {
		foreach my $link (split m/,/ms, $node->link) {
			if (none {
				($_->[0] eq $node->id && $_->[1] eq $link) 
				|| 
				($_->[0] eq $link && $_->[1] eq $node->id)
				} @processed) {

				# Skip link to myself.
				my $link_node = $node_cache_hr->{$link};
				if ($self->{'callback_node_id'}->($self, $node)
					ne $self->{'callback_node_id'}
					->($self, $link_node)) {

					$self->{'callback_edge'}->($self, $node,
						$link_node);
				}
				push @processed, [$node->id, $link];
			}
		}
	}
	eval {
		$self->{'g'}->run(
			'driver' => $self->{'driver'},
			'format' => $self->{'output'},
			'output_file' => $output_file,
		);
	};
	if ($EVAL_ERROR) {
		err 'Cannot create GraphViz output.',
			'Error', $EVAL_ERROR,
			'Dot input', $self->{'g'}->dot_input;
	}

	return;
}

1;

__END__

=encoding utf8

=head1 NAME

Map::Tube::GraphViz - GraphViz output for Map::Tube.

=head1 SYNOPSIS

 use Map::Tube::GraphViz;

 my $obj = Map::Tube::GraphViz->new(%params);
 $obj->graph($output_file);

=head1 METHODS

=over 8

=item C<new(%params)>

 Constructor.

=over 8

=item * C<callback_edge>

 Edge callback.
 Default value is this:
 sub { 
         my ($self, $from, $to) = @_;
         $self->{'g'}->add_edge(
         	'from' => $from,
         	'to' => $to,
         );
         return;
 }

=item * C<callback_node>

 Node callback.
 Default value is \&Map::Tube::GraphViz::Utils::node_color.

=item * C<driver>

 GraphViz2 driver.
 Default value is 'neato'.

=item * C<g>

 GraphViz2 object.
 Parameters 'g' and 'name' cannot combine.
 Default value is this:
 GraphViz2->new(
 	'global' => {
 		'directed' => 0,
 	},
 	$name ? (
 		'graph' => {
 			'label' => $name,
 			'labelloc' => 'top',
 		},
 	) : (),
 );

=item * C<name>

 Name of map.
 Parameters 'g' and 'name' cannot combine.
 Default value is Map::Tube->name or undef.

=item * C<output>

 GraphViz2 output.
 It is required.
 Default value is 'png'.
 Possible values are every formats supported by GraphViz2 module.
 See L<http://www.graphviz.org/content/output-formats>.

=item * C<tube>

 Map::Tube object.
 It is required.
 Default value is undef.

=back

=item C<graph($output_file)>
 
 Get graph and save it to $output_file file.
 Returns undef.

=back

=head1 ERRORS

 new():
         Cannot load UNIVERSAL::DOES module.
         Parameter 'tube' is required.
         Parameter 'tube' must be 'Map::Tube' object.
         Parameter 'output' is required.
         Unsupported 'output' parameter '%s'.
         From Map::Tube::GraphViz::Utils::color_line():
                 No color for line '%s'.
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

=head1 EXAMPLE1

 use strict;
 use warnings;

 use English;
 use Error::Pure qw(err);
 use Map::Tube::GraphViz;

 # Arguments.
 if (@ARGV < 1) {
         print STDERR "Usage: $0 metro\n";
         exit 1;
 }
 my $metro = $ARGV[0];
 
 # Object.
 my $class = 'Map::Tube::'.$metro;
 eval "require $class;";
 if ($EVAL_ERROR) {
         err "Cannot load '$class' class.",
                 'Error', $EVAL_ERROR;
 }
 
 # Metro object.
 my $tube = eval "$class->new";
 if ($EVAL_ERROR) {
         err "Cannot create object for '$class' class.",
                 'Error', $EVAL_ERROR;
 }
 
 # GraphViz object.
 my $g = Map::Tube::GraphViz->new(
         'tube' => $tube,
 );
 
 # Get graph to file.
 $g->graph($metro.'.png');

 # Print file.
 system "ls -l $metro.png";

 # Output without arguments like:
 # Usage: /tmp/SZXfa2g154 metro

 # Output with 'Berlin' argument like:
 # -rw-r--r-- 1 skim skim 1503067 Jan 27 07:24 Berlin.png

=begin html

<a href="https://raw.githubusercontent.com/michal-josef-spacek/Map-Tube-GraphViz/master/images/ex1.png">
  <img src="https://raw.githubusercontent.com/michal-josef-spacek/Map-Tube-GraphViz/master/images/ex1.png" alt="Berlin" width="300px" height="300px" />
</a>

=end html

=head1 EXAMPLE2

 use strict;
 use warnings;

 use English;
 use Error::Pure qw(err);
 use Map::Tube::GraphViz;
 use Map::Tube::GraphViz::Utils qw(node_color_without_label);

 # Arguments.
 if (@ARGV < 1) {
         print STDERR "Usage: $0 metro\n";
         exit 1;
 }
 my $metro = $ARGV[0];
 
 # Object.
 my $class = 'Map::Tube::'.$metro;
 eval "require $class;";
 if ($EVAL_ERROR) {
         err "Cannot load '$class' class.",
                 'Error', $EVAL_ERROR;
 }
 
 # Metro object.
 my $tube = eval "$class->new";
 if ($EVAL_ERROR) {
         err "Cannot create object for '$class' class.",
                 'Error', $EVAL_ERROR;
 }
 
 # GraphViz object.
 my $g = Map::Tube::GraphViz->new(
         'callback_node' => \&node_color_without_label,
         'tube' => $tube,
 );
 
 # Get graph to file.
 $g->graph($metro.'.png');

 # Print file.
 system "ls -l $metro.png";

 # Output without arguments like:
 # Usage: /tmp/SZXfa2g154 metro

 # Output with 'Berlin' argument like:
 # -rw-r--r-- 1 skim skim 885928 Jan 27 07:43 Berlin.png

=begin html

<a href="https://raw.githubusercontent.com/michal-josef-spacek/Map-Tube-GraphViz/master/images/ex2.png">
  <img src="https://raw.githubusercontent.com/michal-josef-spacek/Map-Tube-GraphViz/master/images/ex2.png" alt="Berlin" width="300px" height="300px" />
</a>

=end html

=head1 EXAMPLE3

 use strict;
 use warnings;

 use English;
 use Error::Pure qw(err);
 use GraphViz2;
 use Map::Tube::GraphViz;
 use Map::Tube::GraphViz::Utils qw(node_color_without_label);

 # Arguments.
 if (@ARGV < 1) {
         print STDERR "Usage: $0 metro\n";
         exit 1;
 }
 my $metro = $ARGV[0];
 
 # Object.
 my $class = 'Map::Tube::'.$metro;
 eval "require $class;";
 if ($EVAL_ERROR) {
         err "Cannot load '$class' class.",
                 'Error', $EVAL_ERROR;
 }
 
 # Metro object.
 my $tube = eval "$class->new";
 if ($EVAL_ERROR) {
         err "Cannot create object for '$class' class.",
                 'Error', $EVAL_ERROR;
 }
 
 # GraphViz object.
 my $g = Map::Tube::GraphViz->new(
         'callback_node' => \&node_color_without_label,
         'g' => GraphViz2->new(
                 'global' => {
                         'directed' => 0,
                 },
                 'graph' => {
                         'label' => $metro,
                         'labelloc' => 'top',
                         'overlap' => 0,
                 },
         ),
         'tube' => $tube,
 );
 
 # Get graph to file.
 $g->graph($metro.'.png');

 # Print file.
 system "ls -l $metro.png";

 # Output without arguments like:
 # Usage: /tmp/SZXfa2g154 metro

 # Output with 'Berlin' argument like:
 # -rw-r--r-- 1 skim skim 1212857 Jan 27 07:51 Berlin.png

=begin html

<a href="https://raw.githubusercontent.com/michal-josef-spacek/Map-Tube-GraphViz/master/images/ex3.png">
  <img src="https://raw.githubusercontent.com/michal-josef-spacek/Map-Tube-GraphViz/master/images/ex3.png" alt="Berlin" width="300px" height="300px" />
</a>

=end html

=head1 EXAMPLE4

 use strict;
 use warnings;

 use English;
 use Error::Pure qw(err);
 use Map::Tube::GraphViz;
 use Map::Tube::GraphViz::Utils qw(node_color_id);

 # Arguments.
 if (@ARGV < 1) {
         print STDERR "Usage: $0 metro\n";
         exit 1;
 }
 my $metro = $ARGV[0];
 
 # Object.
 my $class = 'Map::Tube::'.$metro;
 eval "require $class;";
 if ($EVAL_ERROR) {
         err "Cannot load '$class' class.",
                 'Error', $EVAL_ERROR;
 }
 
 # Metro object.
 my $tube = eval "$class->new";
 if ($EVAL_ERROR) {
         err "Cannot create object for '$class' class.",
                 'Error', $EVAL_ERROR;
 }
 
 # GraphViz object.
 my $g = Map::Tube::GraphViz->new(
         'callback_node' => \&node_color_id,
         'tube' => $tube,
 );
 
 # Get graph to file.
 $g->graph($metro.'.png');

 # Print file.
 system "ls -l $metro.png";

 # Output without arguments like:
 # Usage: /tmp/SZXfa2g154 metro

 # Output with 'Berlin' argument like:
 # -rw-r--r-- 1 skim skim 1141071 Feb 24 08:04 Berlin.png

=begin html

<a href="https://raw.githubusercontent.com/michal-josef-spacek/Map-Tube-GraphViz/master/images/ex4.png">
  <img src="https://raw.githubusercontent.com/michal-josef-spacek/Map-Tube-GraphViz/master/images/ex4.png" alt="Berlin" width="300px" height="300px" />
</a>

=end html

=head1 DEPENDENCIES

L<Class::Utils>,
L<English>,
L<Error::Pure>,
L<GraphViz2>,
L<List::MoreUtils>,
L<Map::Tube::GraphViz::Utils>,
L<Scalar::Util>.

=head1 SEE ALSO

=over

=item L<Map::Metro::Graph>

Map::Metro graph.

=item L<Task::Map::Tube>

Install the Map::Tube modules.

=item L<Task::Map::Tube::Metro>

Install the Map::Tube concrete metro modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Map-Tube-GraphViz>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 CONTRIBUTORS

=over

=item Gisbert W. Selke L<gws@cpan.org>

=back

=head1 LICENSE AND COPYRIGHT

© 2014-2023 Michal Josef Špaček

Artistic License

BSD 2-Clause License

=head1 VERSION

0.08

=cut
